### DISCUSSION TITLE
scrolling through favorite filters

### DISCUSSION THREAD

**Discussion by N3C2L** [madprops/grasshopper#51] [Ideas]
I love that you can scroll through the filters by hovering over the filter button on top and using the mouse wheel to scroll the filters. The only negative aspect is, that users can't put their favorite filters in that menu. Such a customization option for this menu would be amazing. If you don't want to touch this menu, maybe implement another button that replaces this button if set in option. Or maybe implement an additional button to the left of the actual filter button (this new button could show only the filter icon inside without the text).

**Comment by madprops** [comment]
Added a `Filter Menu` in `Filter` that appear when `right clicking` the Filter Menu. If menu is empty it shows another thing.

**Comment by N3C2L** [comment]
Thanks! Sorry, I was not clear in my first message above. I meant the other scrolling thing: If you hover the mouse over the filter button on top (don't press left or right click when the mouse cursor is on this button) and then use your mouse wheel to quickly switch between the filters. The only negative aspect is, that users can't put their favorite filters in that menu. Such a customization option for this menu would be amazing. If you don't want to touch this menu, maybe implement another button that replaces this button if set in option. Or maybe implement an additional button to the left of the actual filter button (this new button could show only the filter icon inside without the text).

**Comment by N3C2L** [comment]
I am sorry to reopen this thread, but imagine just to use the mouse wheel to switch trough your favorite filters without a mouse click (see my previous post above). You continued to integrate this switching feature again in the tab box button on top, and I like it. If you could implement such a menu (maybe allow only filters in this menu) I would then be able to switch trough my all my favorite color filters (without clicking or typing, just by using the mouse wheel)

I'd love to be able to customize this top filter menu and use it with the feature described above:
<img width="71" alt="grasshopper filter menu" src="https://github.com/madprops/grasshopper/assets/74585843/bd42f3c4-bd45-4c68-b964-6d6d966fd160">

**Comment by madprops** [comment]
I'm refactoring how the filters work. There will be a new `Cycle Filters` setting too.

**Reply by N3C2L** [reply]
This is truly magnificent, thank you very much for that! Your new implementation certainly makes working with filters easier in many places. I am very excited to be able to switch between these filters more quickly, thanks for the great command at this point.

**Comment by madprops** [comment]
In `v2212` there's the new filter system.
Also new `Cycle Filters` setting.
Added `Custom` to the filter context menu.
Fixed `Custom Filters` which had a bug.
Other stuff.

**Reply by N3C2L** [reply]
Wonderful, thanks.
I've added my favorite filters in the new wheel filter menu.
Cycling through them with the mouse wheel is great, thanks.
A small enhancement could be: right click on the filter button on top shows the filters that where added to the wheel filters menu.

Just for testing purposes I put all the available filters to the wheel menu and found that these aren't functional:
all tags, all colors, single colors, all icons, titled. They show all tabs instead.

**Reply by madprops** [reply]
Yeah I had to fix a bunch of stuff. Uploading in a while.

**Reply by madprops** [reply]
Should be working better now.

**Reply by madprops** [reply]
Also the filter setting label has been corrected to `Cycle Filters` as intended.

**Comment by madprops** [comment]
Ok i'm changing this setting to `favorite_filters`.  Plus they show on right click. Upload soon.

**Comment by N3C2L** [comment]
Absolutely fantastic, and a big thank you. Everything works wonderfully, and I find great joy in using it. The right-click menu at the top is very helpful for direct access to a filter, so thank you for that.

 I noticed 3 interesting things:

 1. It would be another great thing if, with this cycling filter feature, you could also switch to the last active tab per view. For example, when you cycle to "filter red" with the mouse wheel, you would automatically switch to the last red-activated tab. And when you cycle to "filter blue" with the mouse wheel, you would automatically switch to the last blue-activated tab. Maybe an additional option for users to activate unloaded tabs when changing the filter, because they are still included in the recent tabs list.

 2. It's somehow strange that the filter button on top changes its size constantly while cycling. Of course, it makes sense because not every filter title has the same length. But when you end up on the "All" filter, the button has become so small that sometimes you have to target the button again. I would be glad if you could rename that filter to "All Tabs" + an icon (=more length, easier to cycle, bigger button). 

 3. Is it feasible to add another cycling functionality to a different area? I'm considering to cycle the recent tabs: For instance, when hovering over the search bar, using the mouse wheel to navigate tabs; Mouse wheel up = go to the previous tab, Mouses wheel down = go to the next tab.

**Comment by madprops** [comment]
1 sounds complicated

2, I'll add an icon to All

3, what do you mean by search bar, where would the wheel effect be?

Right clicking the Step Back button brings a recent list, and this button can also be clicked to go to previous tabs

Wheel on the text filter input would maybe cycle between used filters

**Reply by N3C2L** [reply]
Ok, thanks.
You are right, the search bar is maybe not the right place for the previous tabs cycle thing.
But maybe in the Tab Box. There could be an option if somebody wants to cycle the previous or next tab with the mouse wheel or cycle between the already existing modes there (pins,header,color,playing,recent).

**Comment by N3C2L** [comment]
The favorite filter cycling has some bugs:
If you cycle them:
1. sometimes the new added single colors have random names and numbers as title, visible when filtered in the top button.
2. sometimes the new added single colors have their previous name displayed in the top filter button.
3. some filters are skipped during the cycling like "all tags" if they are unloaded or sometimes they appear when cycling forward but reappear when cycling backward.

**Reply by madprops** [reply]
1 will get fixed, which I think might fix 2. I don't understand 3, what do you mean by unloaded? They seem to cycle fine

**Reply by N3C2L** [reply]
You already solved all 3 points, thanks. But I think the "All Colors" filter is buggy. Can't cycle to it.

**Comment by madprops** [comment]
>you could also switch to the last active tab per view.

I'm implementing this.

**Reply by N3C2L** [reply]
Big thanks!

**Reply by N3C2L** [reply]
I just tested the sticky filter. Thank you. Would it be possible for the tab to also be automatically activated and not just selected when switching the filter?

**Reply by madprops** [reply]
Yeah now it has 2 modes to choose from, Select, and Activate

**Reply by N3C2L** [reply]
Many thanks again for implementing this! Cycling the views and activate the last items per view; this saves time and clicks.

**Comment by N3C2L** [comment]
There have been 2 issues recently:

1. If "check command" is disabled, cycling or activating other views no longer works.

2. I have the auto-restore function set to 'never' and use the ESC key or the step-back command to return to the 'All' view, which is then like a manual restore. 
However, when the Sticky Filter is activated, the problem is that the 'All' view no longer allows a restore because the last active tab is then activated in the 'All' view. This means I can no longer manually trigger the restore with the ESC key or the step-back command. But I would like to continue using both functions simultaneously for the 'All' view. When cycling through the filters with the mouse wheel, the 'All' view should still have the Sticky Filter active. And if I press the ESC key or the step-back command, the restore function for the 'All' view should be active.

 The problem can be reproduced as follows:
   1. Click on an uncolored tab in the 'All' view.
   2. Color another tab red.
   3. Switch to the red filter.
   4. Click on the red tab.
   5. Press the ESC key or the step-back command.
   6. Problem: The 'All' view does not stay on the red tab but switches to the previous other tab.

**Comment by madprops** [comment]
These should be fixed now.

Also: I changed how the command check works. If the alt is valid it still shows the commands. It checks if command or alt is valid.

**Reply by N3C2L** [reply]
Yeah, thanks that you fixed it!

**Comment by N3C2L** [comment]
I was thinking of a new refined filtering menu. Currently, when you, for example, filter by 'All Colors,' the loaded tabs and the unloaded tabs are always mixed. But what if you could filter them even more precisely? This would be triggered by the middle mouse button on the search input field. In this menu, you could further refine the current filter (All Colors), for example, by 'Loaded,' 'Unloaded,' 'Header'.

**Reply by madprops** [reply]
How about middle clicking items in the filter menu. Like opening the filter menu, then put the cursor on Unloaded and middle click it. It would then perform a filter by considering only the tabs that are already visible

**Reply by N3C2L** [reply]
Yeah, that would be great.

**Reply by N3C2L** [reply]
But I think it would have looked cleaner in an own refine menu, because new user my not know about this feature. And the mouse cursor wouldn't have to move across the other filters that can't be refined with the middle mouse click.

**Reply by madprops** [reply]
How do you envision this? You mentioned "middle mouse button on the search input field" that would also be hard to discover naturally. Also if I re-use the filter menu I have everything available, including submenus like colors

**Reply by N3C2L** [reply]
Ok, you are right.

**Reply by N3C2L** [reply]
But a user couldn't customize their own refine menu. This way, you would have the advantage of having the most used refine filter at the top if set there. Otherwise, in the current menu, you might always have to scroll all the way down to use your preferred refine filter.

**Reply by madprops** [reply]
There's favorite filters. Those could also have the middle click action.

**Reply by N3C2L** [reply]
Yeah, that would be cool, thanks.

**Reply by N3C2L** [reply]
But how about a favorite refine menu where each individual could load only their favorite refines, and by default, it's empty. And in this, you could then load only new refine commands like 'Refine Loaded' or 'Refine Header'.

**Reply by madprops** [reply]
There's too many lists. Custom Filters, Favorite Filters... I don't like that idea
It overlaps with Favorite Filters which appear on right click
Also it's an advanced feature, and just knowing that you can middle click on various filter buttons is probably enough.

**Reply by madprops** [reply]
Also there's already the written concept of "Refine" used elsewhere

**Reply by madprops** [reply]
But then again, adding lists is trivial. I might as well do that middle click thing.

**Reply by N3C2L** [reply]
But the new favorite refine menu could be described in the settings for new users. In the tooltip for that new menu the user could be informed, that this favorite refine menu spawns when middle clicking the search input field. But of course it can be named to something else if you don't want to use "Refine".

**Reply by madprops** [reply]
I added the Refine Filters. The old Refine got renamed to Exact.

**Reply by N3C2L** [reply]
Thank you so much!

**Reply by madprops** [reply]
Alright it's up. Also Exact had a bug I wasn't aware of, that's fixed now

**Reply by madprops** [reply]
Middle click on filter menus work. Refine Filters on middle click Filter works

**Reply by madprops** [reply]
Also added commands to show favorite filters, and refine filters

**Comment by N3C2L** [comment]
Thank you very much for the implementation!

**Comment by N3C2L** [comment]
This idea has been implemented, so I can close this thread.
